include/zephyr/pm/device.h

```diff
@@ -120,6 +120,17 @@ typedef int (*pm_device_action_cb_t)(const struct device *dev,
 typedef bool (*pm_device_action_failed_cb_t)(const struct device *dev,
 					 int err);
 
+/**
+ * @brief PM device reinitialization callback.
+ *
+ * @param dev Device instance.
+ *
+ * @retval 0 If successful.
+ * @retval -ENOTSUP If the requested action is not supported.
+ * @retval Errno Other negative errno on failure.
+ */
+typedef int (*pm_device_reinit_cb_t)(const struct device *dev);
```

Comment:
I think this can be solved in a simpler way:

Use the pm_notifiet_* API (now device agnostic, would require a device context in the callback), so that a driver gets called when the system enters/exits a power state. In these callbacks, driver can perform re-initialization if it knows that it has lost its state.

Reply:
That approach adds more overhead. For every `pm_state` exit, each driver will have to run comparison logic against the current power state and its list of `reinit-power-states`. I've had a similar discussion with @ceolin [here](https://github.com/zephyrproject-rtos/zephyr/pull/59200#discussion_r1263415999).

At least on the STM32 platforms, the SoC will regularly enter and exit low-power modes several dozen times a second, due to the short residency intervals. I think any additional overhead in this case should be avoided. (Adding calculation of exit-latency is another related case).

Reply:
Can overhead be measured?

Assuming I have e.g. 2 devices subscribed, PM will just iterate over 2 list elements and call. Then, the callback will be something like

```
if (state == STOP2) {
   restore_context()
}

return;
```

Subscription could also be made per-state, maybe.

Before adding functionality such as "device reinit", I'd like to understand if this is something common, needed by other vendors to come up with something that is generic enough. For now this seems to be an STM32 requirement, so it can be solved inside drivers, not impacting other PM users.

Reply:
I suppose it adds a few cycles for each driver's notifier call, though it can quickly become a multiple of n, due to the nesting caused by substates (haven't tried to measure it directly). If you have 2 devices subscribed that require reinit in power-states `&stop0`, `&stop1`, and `&stop2`, you now have O(3n) complexity. This is the case on STM32WB for instance. I'm trying to think of and cover the worst case scenarios in my argumentation.

If you take quick look at [STM32WLE5x RM0461, chapter 5.4 - Low-Power Modes, Table 37](https://www.st.com/resource/en/reference_manual/rm0461-stm32wlex-advanced-armbased-32bit-mcus-with-subghz-radio-solution-stmicroelectronics.pdf), you'll see that there are **16(!)** peripherals that require reinit after STOP2 state. For other SoCs like STM32WB, this can start as early as STOP0, leading to a potential worst case of O(16*3) iterations to be run.

> Before adding functionality such as "device reinit", I'd like to understand if this is something common, needed by other vendors to come up with something that is generic enough. For now this seems to be an STM32 requirement, so it can be solved inside drivers, not impacting other PM users.

Very good point. It'd be good to know if this has a potential use case for other platforms. I'll do some cursory research to figure this out.

> Subscription could also be made per-state, maybe.

I also very much like this alternative if expanded to substates. It'd help reduce the O(m) complexity from per-driver-instance to O(m) in total. (where m = reinit substates)

Reply:
> I'm working on a proposal to simplify to vendor-specific states, ie, one single state holder. The current approach of aligning to ACPI lead to a quite cumbersome implementation

Very interesting! I agree, forcing MCUs into the ACPI sleep states pattern makes things unnecessarily difficult—especially when trying to fit the various flat MCU power state designs into substates under these.

Reply:
> Before adding functionality such as "device reinit", I'd like to understand if this is something common, needed by other vendors to come up with something that is generic enough. For now this seems to be an STM32 requirement, so it can be solved inside drivers, not impacting other PM users.

I've checked a few different vendor SoCs that are currently supported in Zephyr, and I've found at least one other that requires reinit functionality: the TI SimpleLink CC13x2/CC26x2 series.

According to the [reference manual](https://www.ti.com/lit/ug/swcu185f/swcu185f.pdf?ts=1689649773475), *chapter 7.6.4 - Standby Mode* (page 519):
*"All modules in MCU_VD with retention, as shown in Figure 7-3, are retained in standby mode. All other logic in MCU_VD must be reconfigured after wake up from standby mode."*
Modules in this case being TI-lingo for peripherals.

If you check *chapter 7.4 Digital Power Partitioning, Figure 7-3* (page 509), you'll see that this encompasses many generic peripherals such as I2C, SSI, etc.

Interestingly enough, the CC13xx/CC26xx I2C driver currently even does [reinit](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/i2c/i2c_cc13xx_cc26xx.c#L294-L319) via [HAL](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/i2c/i2c_cc13xx_cc26xx.c#L373-L375) in a manner very similar to the `pm_notify` callback implementation, but specific to the power state. *(It also does some [strange things](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/i2c/i2c_cc13xx_cc26xx.c#L382-L384) related to power management I don't quite understand)*

Would this make you consider the implementation in this PR? I'd also create a separate PR for the appropriate changes to TI SimpleLink DTS and drivers affected. I still feel the DT property method is a bit more elegant than driver `pm_notify` callbacks.

Reply:
> Regarding state/substate, I'm working on a proposal to simplify to vendor-specific states, ie, one single state holder. The current approach of aligning to ACPI lead to a quite cumbersome implementation, because MCUs have different requirements that in practice do not align or deviate too much.

note that MCUs are not the only one consuming it there are targets where ACPI is more aligned.

Reply:
It'd be nice to have coexistence between ACPI and custom MCU-specific power states. The substates solution is just clunky, IMHO.
After all, ACPI platforms are a minority on Zephyr (do correct me if I'm wrong).

Reply:
> I suppose it adds a few cycles for each driver's notifier call, though it can quickly become a multiple of n, due to the nesting caused by substates (haven't tried to measure it directly). If you have 2 devices subscribed that require reinit in power-states `&stop0`, `&stop1`, and `&stop2`, you now have O(3n) complexity. This is the case on STM32WB for instance. I'm trying to think of and cover the worst case scenarios in my argumentation.
> 
> If you take quick look at [STM32WLE5x RM0461, chapter 5.4 - Low-Power Modes, Table 37](https://www.st.com/resource/en/reference_manual/rm0461-stm32wlex-advanced-armbased-32bit-mcus-with-subghz-radio-solution-stmicroelectronics.pdf), you'll see that there are **16(!)** peripherals that require reinit after STOP2 state. For other SoCs like STM32WB, this can start as early as STOP0, leading to a potential worst case of O(16*3) iterations to be run.
> 

That is not really the case right ? We don't transition from stop0 -> stop1 -> stop2, states change always from active to some of these power states. So it is not like you will be called 16  * 3 when going to sleep. The problem though is when you just need to reinit in stop2 but you will be called when going to to stop0 or stop1.

> > Before adding functionality such as "device reinit", I'd like to understand if this is something common, needed by other vendors to come up with something that is generic enough. For now this seems to be an STM32 requirement, so it can be solved inside drivers, not impacting other PM users.
> 
> Very good point. It'd be good to know if this has a potential use case for other platforms. I'll do some cursory research to figure this out.
> 
> > Subscription could also be made per-state, maybe.
> 
> I also very much like this alternative if expanded to substates. It'd help reduce the O(m) complexity from per-driver-instance to O(m) in total. (where m = reinit substates)

That is something that makes sense to me too.

Reply:
The property makes sense to me, what I am not so sure is the additional callback, I would like to explore a little bit more the notifier path. It seems to me that we can make that work for this.

Reply:
> That is not really the case right ? So it is not like you will be called 16 * 3 when going to sleep

I meant in the case of a node having 3 substates in `reinit-power-states`, specifically with the pm_notifier callback only being aware of only the main `pm_state` (no substate-specific cb). In that case the driver callback comparison logic would have to check if the exit substate matches any of the three, thus in the worst case iterating three times before it hits a match on STOP2, for instance.

Also forgot to mentioned that of the 16 peripherals some have more than one instance, e.g. I2C-1 and -2, so worst-case would be more than 16 driver instance pm_notifier callbacks getting called (around 24ish).

If the two of you are still in favor of a pm_notifier approach over my approach in this PR:
Ideally we should modify the pm_notifier callback registration to behave more like the [TI SimpleLink HAL notifier registration](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/i2c/i2c_cc13xx_cc26xx.c#L373-L375), i.e. the callback gets called only when it matches the exact pm (sub)state entrance/exit.

Just give me the go-ahead that'd you'd both prefer this approach, and I'll look to implement it in place of the solution proposed in this PR.

Reply:
> Just give me the go-ahead that'd you'd both prefer this approach, and I'll look to implement it in place of the solution proposed in this PR.

> I would like to explore a little bit more the notifier path. It seems to me that we can make that work for this.

I'll take that as a green light from you :)